Lookup properties through label

toban opened the pull request.

lucaswerkmeister commented on includes/ServiceWiring.php:
```diff
@@ -53,7 +58,52 @@
-			WikibaseReconcileEditServices::getReconciliationService( $services )
+			WikibaseReconcileEditServices::getReconciliationService( $services ),
+			WikibaseReconcileEditServices::getPropertyLabelResolver( $services ),
+			LoggerFactory::getInstance( 'WikibaseReconcileEdit.MinimalItemInput' )
```
This seems too specific to me – each log channel has to be configured in the site config, after all. `'WikibaseReconcileEdit'` should be enough, I think.

guergana replied:
Updated!

lucaswerkmeister commented on includes/InputToEntity/MinimalItemInput.php:
```diff
@@ -113,6 +130,45 @@ public function getItem( array $inputEntity, PropertyId $reconcileUrlProperty )
+			try {
+				return new PropertyId( $propertyId );
+			} catch ( \InvalidArgumentException $exception ) {
+				$this->logger->warning( 'Tried parsing something weird' );
```
I don’t think this warrants a warning – it’s totally reasonable for a property label to start with `P` (Wikidata has [328 such properties](https://query.wikidata.org/#SELECT%20%2a%20WHERE%20%7B%0A%20%20%3Fp%20a%20wikibase%3AProperty%3B%0A%20%20%20%20%20rdfs%3Alabel%20%3FpLabel.%0A%20%20FILTER%28LANG%28%3FpLabel%29%20%3D%20%22en%22%29%0A%20%20FILTER%28STRSTARTS%28%3FpLabel%2C%20%22P%22%29%29%0A%7D) just in English). I think we should try to parse any string as a property ID first, and if that fails, then fall back to interpreting the string as a label, with at most an `->info()` log message (or maybe even just `->debug()`).

guergana replied:
And after interpreting it as a label do we just use the label in the info or debug message?

guergana replied:
This has been addressed. Tho I am not sure if what I did is what you meant. Could you check again?

lucaswerkmeister replied:
Yup, that looks good now. Though I feel like we could even remove the logger call completely, and then stop injecting a logger and save ourselves a bunch of code. WDYT?

guergana replied:
:shrug: ok!

lucaswerkmeister reviewed (commented):
Some early comments. Overall, this looks pretty good to me, though the copied `PropertyLabelResolver` service wiring is perhaps unsatisfying.

lucaswerkmeister commented on includes/InputToEntity/MinimalItemInput.php:
```diff
@@ -113,6 +130,43 @@ public function getItem( array $inputEntity, PropertyId $reconcileUrlProperty )
+		$entityIds = $this->propertyLabelResolver->getPropertyIdsForLabels( [ $label ] );
+
+		if ( empty( $entityIds ) ) {
+			throw new PropertyLabelNotResolvedException( $label, 'en' );
```
That’s a WikibaseClient exception, and it also uses a message key that only exists in client. If we want to depend on Repo only, I think we should throw a `ReconciliationException` here, with a message of our own.

guergana replied:
This seems resolved.

lucaswerkmeister commented:
Open TODOs I remember:

- add tests in `MinimalItemInputTest`
- look into the `mockPropertyLabelResolver()` in `MinimalItemInputTest` and `EditRequestParserTest`, those look weird to me

guergana commented on includes/ServiceWiring.php:
```diff
@@ -9,7 +9,11 @@
 use MediaWiki\MediaWikiServices;
+use Wikibase\DataModel\Entity\Property;
 use Wikibase\DataModel\Services\Statement\GuidGenerator;
+use Wikibase\Lib\Store\Sql\Terms\CachedDatabasePropertyLabelResolver;
```
@toban and I worked on this and we have some doubts/comments: The `CachedDatabasePropertyLabelResolver` is not yet working with Federated Properties. We tested it. We could build our own or make it work with wikibase. `CachedDatabasePropertyLabelResolver` is only used in client. What approach do you think we should take @lucaswerkmeister

lucaswerkmeister replied:
Let’s merge this without FP support and then see if we need it later.

guergana replied:
Sounds ok to me :)

lucaswerkmeister commented on i18n/qqq.json:
```diff
@@ -18,5 +18,6 @@
 	"wikibasereconcileedit-editendpoint-invalid-reconciliation-statement-type": "Error returned by the API if the “entity” request parameter if the reconciliation statement's mainsnak is not of type value",
 	"wikibasereconcileedit-reconciliationservice-matched-multiple-items": "Error returned by the API if the reconciliation request matches multiple items.\n\nParameters:\n* $1 – the matched items (comma-separated list)\n* $2 – the number of matched items",
 	"wikibasereconcileedit-minimaliteminput-required-keys": "Error returned by the API if the minimal item input statements parameter is missing a certain key.\n\nParameters:\n $1 - the missing key",
+	"wikibasereconcileedit-minimaliteminput-found-multiple-properties" : "Error returned by the API if the minimal item input result found multiple matching properties with the same label.\n\nParameters:\n $1 - the property ids",
```
This message needs another parameter with the number of properties, since some languages have different plural forms depending on the number.

lucaswerkmeister replied:
(If we keep the message at all, that is. See below.)

toban replied:
removed

lucaswerkmeister commented on includes/InputToEntity/MinimalItemInput.php:
```diff
@@ -148,12 +148,13 @@ public function getPropertyIdByLabel( string $label ): PropertyId {
 		if ( count( $entityIds ) !== 1 ) {
-			throw new \InvalidArgumentException( "Found multiple properties" );
+			throw new ReconciliationException(
+				MessageValue::new( 'wikibasereconcileedit-minimaliteminput-found-multiple-properties' )
```
Hm, didn’t we say in another call that this didn’t need a translatable message since it should never happen?

I think we should remove this check completely, in fact. I think there should just be one condition: `isset( $entityIds[ $label ] )` – if yes, then return it, otherwise “property not found”. We don’t need to care if there are any other entries in the array. (On the other hand, currently we don’t actually check that `$entityIds[ $label ]` is set – we just assume that if the array has one element, it’ll be that one.)

toban replied:
all good points. We were talking about this but couldn't point to where this uniqueness guarantee(property label, language) comes from in wikibase

toban replied:
removed

toban commented on includes/ServiceWiring.php:
```diff
@@ -53,7 +57,52 @@
+		$repo = WikibaseRepo::getDefaultInstance();
+		$settings = $repo->getSettings();
+
+		$languageCode = $services->getContentLanguage()->getCode();
```
Maybe this should be a setting for the WikibaseReconcileEdit extension rather than defaulting to the language of the mediawiki installation?

toban replied:
lets not

toban commented on tests/api-testing/REST/EditEndpoint.js:
```diff
@@ -103,4 +104,58 @@ describe( 'POST /edit', () => {
+		const { status, body } = await client.post(
+			'/edit',
+			params,
+			{ 'content-type': 'application/x-www-form-urlencoded' } // why?
```
this would crash in CI now, needs a rebase

toban replied:
done

toban committed:
Lookup properties through their labels

toban commented:
LGTM!

guergana committed:
Removed strange double variable assignment

lucaswerkmeister committed:
Don’t mock callback in mockPropertyLabelResolver()

These mocks aren’t meant to actually be called: if they’re called, it’s
because the property string doesn’t look like a property ID, but in that
case the callback just would’ve thrown an exception anyways.

lucaswerkmeister reviewed (approved):
Added a minor improvement, looks good to me otherwise. Should be good to squash-merge :+1:
